core/src/main/java/com/linecorp/armeria/common/RpcResponse.java

```diff
@@ -61,6 +63,29 @@ static RpcResponse from(CompletionStage<?> stage) {
         return res;
     }
 
+    /**
+     * Creates a new {@link RpcResponse} that delegates to the {@link RpcResponse} produced by the specified
+     * {@link CompletionStage}. If the specified {@link CompletionStage} fails, the returned response will be
+     * closed with the same cause as well.
+     */
+    static RpcResponse fromResponseFuture(CompletionStage<? extends RpcResponse> stage) {
+        requireNonNull(stage, "stage");
+        final DefaultRpcResponse res = new DefaultRpcResponse();
+        stage.whenComplete((delegate, thrown) -> {
+            if (thrown != null) {
+                res.completeExceptionally(Exceptions.peel(thrown));
+            } else if (delegate == null) {
+                res.completeExceptionally(
+                        new NullPointerException("delegate stage produced a null response: " + stage));
+            } else if (delegate.isCompletedExceptionally()) {
+                res.completeExceptionally(delegate.cause());
+            } else {
+                res.complete(delegate.join());
```

Comment:
This line is dangerous, need to wait delegate completes, so we should do nested whenComplete.